packages/aft/lib/src/changelog/changelog.dart

```diff
@@ -0,0 +1,165 @@
+
+import 'package:aft/src/changelog/commit_message.dart';
+import 'package:aft/src/changelog/printer.dart';
+import 'package:aws_common/aws_common.dart';
+import 'package:built_collection/built_collection.dart';
+import 'package:built_value/built_value.dart';
+import 'package:collection/collection.dart';
+import 'package:markdown/markdown.dart';
+import 'package:pub_semver/pub_semver.dart';
+
+part 'parser.dart';
+part 'changelog.g.dart';
+
+/// Marker version for the `NEXT` (unreleased) version.
+final Version nextVersion = Version(0, 0, 0, pre: nextVersionTag);
+
+/// Tag for the `NEXT` version.
+const String nextVersionTag = 'NEXT';
+
+/// {@template aft.changelog.changelog}
+/// A Dart representation of a `CHANGELOG.md` file.
+/// {@endtemplate}
+abstract class Changelog implements Built<Changelog, ChangelogBuilder> {
+  /// {@macro aft.changelog.changelog}
+  factory Changelog([void Function(ChangelogBuilder) updates]) = _$Changelog;
+  Changelog._();
+
+  /// Parses [changelogMd] for a list of the versions.
+  ///
+  /// Throws a [ChangelogParseException] if there are issues processing the
+  /// changelog.
+  factory Changelog.parse(
+    String changelogMd, {
+    AWSLogger? logger,
+  }) {
+    final parser = Document();
+    final lines = LineSplitter.split(changelogMd).toList();
+    final ast = parser.parseLines(lines);
+    final visitor = _ChangelogParser(logger);
+    for (final node in ast) {
+      node.accept(visitor);
+    }
+    return (visitor.builder..originalText = changelogMd).build();
+  }
+
+  /// The original CHANGELOG.md text.
+  String get originalText;
+
+  /// A map of semantic versions to their nodes.
+  BuiltListMultimap<Version, Node> get versions;
+
+  /// The latest version in the changelog, or `null` if the changelog is empty.
+  Version? get latestVersion => maxBy(versions.keys, (v) => v);
+
+  /// Whether there's a `NEXT` entry in the changelog.
+  bool get hasNextEntry => versions.keys.any((v) => v == nextVersion);
+
+  /// Creates a version entry which can be rendered as markdown for the given
+  /// list of [commits].
+  ///
+  /// If [version] is not specified, it defaults to [nextVersion]. If [version]
+  /// already exists in the changelog, it is updated with the new list of
+  /// commits.
+  List<Node> makeVersionEntry({
+    required Iterable<CommitMessage> commits,
+    Version? version,
+  }) {
+    version ??= nextVersion;
+    commits = commits.where((commit) => commit.includeInChangelog);
+    final commitsByType =
+        commits.groupListsBy<CommitTypeGroup>((element) => element.group);
+
+    final versionText =
+        version == nextVersion ? nextVersionTag : version.toString();
+    final header = Element.text('h2', versionText);
+    final nodes = <Node>[header];
+
+    if (commits.isEmpty) {
+      // If there are no commits worth including, add a generic message about
+      // bug fixes/improvements.
+      nodes.add(Element.text('li', 'Minor bug fixes and improvements\n'));
```

Comment:
In reality, this function was only ever being called in scenario 1. I've added an assert statement and updated the call sites to make it more clear.

Reply:
Ah, yeah whoops that's true. Still, scenario 2 never happens since a bump to component packages includes the commit. I think in the case you point out, we can either add the minor fixes or just not update the changelog. I'm cool with either approach.